pkg/template/servicebroker/servicebroker.go

```diff
@@ -1,62 +1,90 @@
+	kc                 kclientset.Interface
 	templateclient     internalversiontemplate.TemplateInterface
 	restconfig         restclient.Config
-	lister             cache.StoreToTemplateLister
+	lister             templatelister.TemplateLister
 	templateNamespaces map[string]struct{}
+	ready              chan struct{}
 }
 
-func NewBroker(restconfig restclient.Config, localSAR client.LocalSubjectAccessReviewsNamespacer, secretsGetter internalversion.SecretsGetter, informers shared.InformerFactory, namespaces []string) *Broker {
+func NewBroker(privrestconfig restclient.Config, privkc kclientset.Interface, infraNamespace string, informer templateinformer.TemplateInformer, namespaces []string) *Broker {
 	templateNamespaces := map[string]struct{}{}
 	for _, namespace := range namespaces {
 		templateNamespaces[namespace] = struct{}{}
 	}
 
-	return &Broker{
-		secretsGetter:      secretsGetter,
-		localSAR:           localSAR,
-		templateclient:     templateclientset.NewForConfigOrDie(&restconfig).Template(),
-		restconfig:         restconfig,
-		lister:             informers.Templates().Lister(),
+	b := &Broker{
+		lister:             informer.Lister(),
 		templateNamespaces: templateNamespaces,
+		ready:              make(chan struct{}),
 	}
-}
 
-func (b *Broker) getClientsForUsername(username string) (kclientset.Interface, client.Interface, templateclientset.Interface, error) {
-	u := &user.DefaultInfo{Name: username}
+	go func() {
+		// the broker is initialised asynchronously because fetching the service
+		// account token requires the main API server to be running
 
-	oc, err := authclient.NewImpersonatingOpenShiftClient(u, b.restconfig)
-	if err != nil {
-		return nil, nil, nil, err
-	}
+		glog.Infof("Template service broker: waiting for authentication token")
 
-	kc, err := authclient.NewImpersonatingKubernetesClientset(u, b.restconfig)
-	if err != nil {
-		return nil, nil, nil, err
-	}
+		restconfig, oc, kc, _, err := serviceaccounts.Clients(
+			privrestconfig,
+			&serviceaccounts.ClientLookupTokenRetriever{Client: privkc},
+			infraNamespace,
+			bootstrappolicy.InfraTemplateServiceBrokerServiceAccountName,
+		)
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("Template service broker: failed to initialize: %v", err))
+			return
+		}
 
-	impersonatingConfig := authclient.NewImpersonatingConfig(u, b.restconfig)
-	templateclient, err := templateclientset.NewForConfig(&impersonatingConfig)
-	if err != nil {
-		return nil, nil, nil, err
-	}
+		b.oc = oc
+		b.kc = kc
+		b.templateclient = templateclientset.NewForConfigOrDie(restconfig).Template()
+
+		glog.Infof("Template service broker: waiting for informer sync")
 
-	return kc, oc, templateclient, nil
+		for !informer.Informer().HasSynced() {
+			time.Sleep(100 * time.Millisecond)
+		}
+
+		glog.Infof("Template service broker: ready")
+
+		close(b.ready)
+	}()
+
+	return b
+}
+
+func (b *Broker) WaitForReady() error {
+	timer := time.NewTimer(10 * time.Second)
```

Comment:
Any incoming TSB API request would be held up for maximally that length of time before giving up if the broker hasn't initialised.  Initialisation is a one-off event at startup and should be rapid.  10s is in use elsewhere in our codebase in similar scenarios.  At the moment we return status code 500 if the timer expires, arguably that should be a different code, but there's nothing to cover this in the OSB spec and I'm not aware of a specific standardised "try again later" status code to use.

I think I should add some code documentation here :)